### PR TITLE
MQE: Recover from query-evaluation panics caused by invalid stored data

### PR DESCRIPTION
**Pull request by zenador**
#### What this PR does

Previously, when MQE evaluates a query or recording/alerting rule over bad data, e.g. an invalid native histogram with a negative-offset span resulting from a bug in Mimir 2.14-2.17 (see https://github.com/grafana/mimir/issues/16343), the process crashes. A single corrupt series can therefore crash-loop a querier or ruler-querier shared by other tenants.

For the above example, it first started failing when https://github.com/prometheus/prometheus/pull/17561 was vendored into Mimir. It panics when failing validation, but in Prometheus those panics are recovered and only surfaced as query errors, whereas MQE does not recover those panics. Since it generally prefers to return explicit errors instead of using panics for control flow, it never installed a Prometheus-style recover, and its existing recover deliberately re-panicked to crash (https://github.com/grafana/mimir/pull/15753). That's fine for MQE's own code, but not for panics from a vendored library like `model/histogram`.

This PR makes MQE recover panics during evaluation and returns them as a query error, like the Prometheus engine does, so one bad query or series cannot crash a shared querier or ruler. A new advanced flag, `-querier.mimir-query-engine.surface-evaluation-panics` (default `false`), re-raises panics to crash instead. Enable it in testing environments so bugs still fail fast before production.

Recovered panics are counted by the new metric `cortex_mimir_query_engine_evaluation_panics_total`, labelled by `user` and `reason`: `invalid_data` (invalid stored data), `runtime_error` (a Go runtime error, almost certainly a bug), or `other`. `invalid_data` is logged at warning without a stack trace, since it recurs over the same series; the rest are logged at error with the stack trace. A new `MimirQueryEngineEvaluationPanics` alert (warning) and runbook entry cover them.

##### Other approaches considered and discarded

- Repair the invalid spans on read (even if it could be done safely, that change belongs upstream)
- Recover only the negative-offset panic: rejected as it's not future-proof and the next data-triggered library panic could crash in production
- Crash based on panic type (kept as the `reason` label instead, since even an MQE assertion on production data should not crash a shared process)
- Per-call-site wrappers around every histogram operation: rejected as it's more code and no advantage over a single boundary recover that discriminates by the panic value's type

#### Which issue(s) this PR fixes or relates to

Loose follow up to https://github.com/grafana/mimir/pull/15753
Partially addresses https://github.com/grafana/mimir/issues/16343

#### Checklist

- [x] Tests updated.
- [x] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`. If changelog entry is not needed, please add the `changelog-not-needed` label to the PR.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/streamingpromql/evaluator.go [hunk 1]
```diff
@@ -91,15 +94,74 @@ func (e *Evaluator) Evaluate(ctx context.Context, observer EvaluationObserver) (
...
+	// reliably scraped, and the crash is visible on its own.
 	defer func() {
-		if r := recover(); r != nil {
+		r := recover()
+		if r == nil {
+			return
+		}
+
+		if _, isRuntimeErr := r.(runtime.Error); isRuntimeErr {
+			level.Error(logger).Log("msg", "runtime error while evaluating query, re-panicking to crash", "err", r, "expr", e.originalExpression)
```
[the remark this answers is not in the document]
Comment (zenador) [reply pkg/streamingpromql/evaluator.go:121]: Tested with Claude: same-schema `Add`/`Sub`/`KahanAdd`, the rate `Sub` path over two same-layout corrupt points, `DetectReset`, and full bucket-iterator walks over negative-offset-span histograms (with both zero and nonzero counts in the misplaced bucket) complete without any panic — the same-schema paths don't validate the layout at all. The only panics this data produces come from the resolution-reduction family, which panic with `histogram.Error` and are recovered by this PR. Index-out-of-range would require a span/bucket-count mismatch, which the chunk format can't represent, so it isn't reachable from stored data. The residual point — a hypothetical future `runtime.Error` from data would still crash — is the deliberate fail-fast trade-off described in the PR, not an unhandled case of this bug.

[the remark this answers is not in the document]
Comment (zenador) [reply pkg/streamingpromql/evaluator.go:121]: The original frames aren't lost: deferred functions run before the stack unwinds, so re-panicking from the recover preserves them, and the crash traceback shows the full chain to the panic site marked `[recovered, repanicked]` (verified — the pre-fix crash dumps for #16343 include the `mustReduceResolution` → `Add` → operator frames below the handler). No stack capture is needed on the re-panic path; the fatal dump already contains it.

**Comment by tcp13equals2** [comment]
So (according to Claude) there are ~67 panic( call sites in non-test pkg/streamingpromql code) and those sampled are not runtime errors. 

Is it expected that all of these would now be recovered? 

Would it be safer to start with cherry picking the panics we know are safe to recover from (such as the histogram errors) and then build out a list of other MQE panics we want to recover from?

`Genuine runtime errors are still re-panicked so engine bugs remain visible` is not really true, as all these existing panics will not be re-panicked.

**Comment by zenador** [comment]
@tcp13equals2 Thanks for the review! (And sorry for long reply)

> Is it expected that all of these would now be recovered?

Yes, in this PR my intention is to make most of the MQE panics (except for runtime.Error) recover, because this is the same as how Prometheus engine handles it, and because in production we'd much rather that MQE panics fail queries than take down the entire querier or ruler-querier, especially in a multi-tenant setup, and the retrying of queries could shut down other queriers as well. The way I see it, the way we can panic in MQE currently is an oversight, because we handle most places Prometheus panics with explicit errors, but that doesn't save us from panics happening within Prometheus code.

> Would it be safer to start with cherry picking the panics we know are safe to recover from (such as the histogram errors) and then build out a list of other MQE panics we want to recover from?

We could do that if we just wanted to fix the original problem that prompted this, but with the allowlist approach, if we miss something now, or later on fail to account for a new potential panic in Prometheus when vendoring, it could cause an incident in future. So I guess the question is what "safe" means, are we trying to prevent production incidents, or are we trying to make sure we fail visibly so we can catch it earlier? The only panics we don't want to recover from are those that will always trigger reliably and not in some weird edge-case, so that we are 100% sure we can catch them in test environments before going out to prod, but we might not be able to easily distinguish those. I'm worried about new subtle write bugs that produce bad data like what prompted this issue originally, because we might not see such data in our testing environments.

In this PR we recover from panics but still make them visible through a metric, which we will have to track to make sure we don't silently introduce new problems. Adjusting to a new way failures could surface could be tricky, but we could also argue that this metric could be a better way to surface potential problems once we adjust to start tracking it properly, because they are limited to panics, and sometimes we might not notice a simple restart of some queriers as that could also be triggered by other conditions like OOM. Additionally, along with the metric, the panics most likely to indicate a bug (`reason="other"`) are logged at error level with a full stack trace, so unlike a restart they point loudly and unambiguously at what needs investigation. Known data-validation errors (`reason="invalid_data"`) are logged at warning level without a stack, since their origin is understood and they can recur on every evaluation over the same bad series.

Alternatively, I also considered but discarded the idea of making the re-panics/recovers conditional on a config. So we would set the default to recover everything so that in production we don't panic, and in our testing environments we can override that to re-panic everything. Initially I thought it was a bit janky, but it does have its benefits. In testing environments we could still notice all the panics, and we wouldn't crash production queriers. We could get rid of the distinction between runtime.Error and others. However, I'd still like to have the new metric and alerts so we still have a way to track these panics in production if it's triggered by weird data we don't have in our testing environments. So choosing this approach wouldn't really simplify anything, and would add a new config, but I'm open to this if we prefer panicking as much as possible in test.

> `Genuine runtime errors are still re-panicked so engine bugs remain visible` is not really true, as all these existing panics will not be re-panicked.

Good point, thanks for catching that! Will fix the descriptions in the PR and changelog when we settle on an approach to take. If we want to ensure that we re-panic those, one thing we can do is check the stack origin, and re-panic only if the package prefix matches Mimir (so it's not from Prometheus code). This is just an option, I think it would be cleaner to have consistent behaviour regardless of origin, but it could always be added as a metric label.

**Comment by 56quarters** [comment]
> Yes, in this PR my intention is to make most of the MQE panics (except for runtime.Error) recover, because this is the same as how Prometheus engine handles it, and because in production we'd much rather that MQE panics fail queries than take down the entire querier or ruler-querier, especially in a multi-tenant setup, and the retrying of queries could shut down other queriers as well. The way I see it, the way we can panic in MQE currently is an oversight, because we handle most places Prometheus panics with explicit errors, but that doesn't save us from panics happening within Prometheus code.

I don't agree that we should recover from places where MQE panics explicitly. We have control of the MQE code and any place we are panicking today could be changed to an error instead. I think recovering panics makes sense for the original motivating case (invalid histogram data that we can't do anything about).

**Comment by zenador** [comment]
@56quarters As mentioned on the internal thread, we're trying for a recover everything in prod (with metrics and logging with stack trace), continue panicking everything in test environments approach, so we can treat explicit MQE panics the same as the rest (since we still wouldn't want to crash prod instances, especially if shared by multiple tenants) and still get the same visibility as before in test environments. Would that be okay?

**Review by JStickler — COMMENTED** [review]
[docs team]  Docs LGTM

File: pkg/streamingpromql/evaluator.go
Comment (karlskewes) [inline pkg/streamingpromql/evaluator.go:54]: This function is getting difficult for me to deduce correct ordering in part due to large deferred functions. 

I propose we split out to something like the following, which also enables cheap `classifyPanic` unit tests:

```go
// logEvaluationStats logs the query's evaluation stats and records its peak memory consumption.
func (e *Evaluator) logEvaluationStats(logger *spanlogger.SpanLogger, rootQueryID string, err error)

// handleEvaluationPanic recovers a panic raised while evaluating the query. It must be the last
// defer registered by Evaluate, so it runs before the cleanup defers and never sees a panic they
// raised: a pool or memory-tracker invariant violation during cleanup must crash the process,
// because the pools are shared between tenants.
func (e *Evaluator) handleEvaluationPanic(ctx context.Context, logger *spanlogger.SpanLogger, err *error)

// classifyPanic returns the value for the reason label of cortex_mimir_query_engine_evaluation_panics_total.
func classifyPanic(r any) string

// logPanicWithStack logs r with the current stack trace. It must be called while the panic is
// unwinding, so the trace still reaches the panic site.
func logPanicWithStack(l log.Logger, msg string, r any, expr string)
```

File: pkg/streamingpromql/evaluator.go [hunk 2]
```diff
@@ -99,15 +103,76 @@ func (e *Evaluator) Evaluate(ctx context.Context, observer EvaluationObserver) (
...
+	//
+	// surfaceEvaluationPanics decides what happens, regardless of the panic's source:
+	//   - Enabled (dev and ops): re-raise every panic so it crashes the process and bugs fail fast.
+	//   - Disabled (the default, production): convert every panic into a query error, matching the
+	//     Prometheus engine, so one bad query or series cannot take down the component.
+	//
+	// Either way the stats logging above reports the query as failed. Recovered panics are counted,
+	// labelled by tenant and a coarse reason, and logged with a stack trace (except known invalid-data
+	// panics, which can recur on every evaluation over the same series).
 	defer func() {
```
Comment (karlskewes) [inline pkg/streamingpromql/evaluator.go:119]: The panic handler is registered before the defers that close the operators.
Deferred calls run last-registered-first, so the handler runs after them.
It therefore covers panics raised by the cleanup, not only by the evaluation.

That was harmless before this PR.
The old handler recovered cleanup panics too, but re-panicked unconditionally, so they always crashed the process.
With `surface_evaluation_panics` disabled they now become query errors instead.

- `MemoryConsumptionTracker.DecreaseMemoryConsumption` panics when a slice is returned to a pool twice (`pkg/util/limiter/memory_consumption.go:419`).
  It is raised from `Put`, inside operator `Close()`, inside `closeOperators()`.
  The handler catches it, counts it as `reason="other"`, and the process keeps serving from a pool it was just told is corrupt.
  The pools are process-global and shared between tenants, so this case must crash.
- `recover()` returns the cleanup panic, not the one already unwinding.
  The returned error and the `reason` label then describe the cleanup rather than the original bug.
  The logged stack trace still holds both sites, so this one is a degradation, not a loss.

Registering the handler last makes it the first deferred call to run while the stack unwinds.
It then sees only panics from the evaluation.
Panics from the cleanup escape and crash in both settings of the flag.

The one cost is that `activeQueryTracker.InsertWithDetails` is no longer covered.
No placement can both run before the cleanup and cover that call.
It builds a description string and touches no pool, so I think that is an acceptable trade.

Reasoning about any of this needs the defer registration order to be visible, which is easier after the extraction in https://github.com/grafana/mimir/pull/16383/changes#r4078918621.

File: pkg/streamingpromql/evaluator.go [hunk 3]
```diff
@@ -99,15 +103,76 @@ func (e *Evaluator) Evaluate(ctx context.Context, observer EvaluationObserver) (
...
+
+		// Classify the panic so data problems and likely bugs can be told apart without reading logs.
+		// Validation errors from the histogram library mean invalid stored data; a Go runtime error is
+		// almost certainly an engine bug; anything else is unclassified and may also be a bug.
+		_, isRuntimeErr := r.(runtime.Error)
+		reason := "other"
+		if isErr {
+			var validationErr histogram.Error
+			switch {
+			case errors.As(rErr, &validationErr):
```
Comment (karlskewes) [inline pkg/streamingpromql/evaluator.go:161]: nit: would be nice to use the standard library `errors` instead of github.com/pkg/errors. It's a straight swap in imports.
There are only a couple of usages in this file.

File: pkg/streamingpromql/evaluator.go [hunk 4]
```diff
@@ -99,15 +103,76 @@ func (e *Evaluator) Evaluate(ctx context.Context, observer EvaluationObserver) (
...
+		userID := ""
+		if tenantIDs, tenantErr := tenant.TenantIDs(ctx); tenantErr == nil {
+			userID = tenant.JoinTenantIDs(tenantIDs)
+		}
+
+		// Classify the panic so data problems and likely bugs can be told apart without reading logs.
+		// Validation errors from the histogram library mean invalid stored data; a Go runtime error is
+		// almost certainly an engine bug; anything else is unclassified and may also be a bug.
+		_, isRuntimeErr := r.(runtime.Error)
+		reason := "other"
```
Comment (karlskewes) [inline pkg/streamingpromql/evaluator.go:157]: nit: `other` might be slightly off. 

It reads as "not invalid data, therefore probably a bug", and the runbook follows that at `mimir-runbooks/_index.md:402` with "Report the bug". 

But only 1 panic site in `model/histogram` can carry a `histogram.Error`, see: `panic(err)` in `mustReduceResolution` (`generic.go:943`).
 
The other 16 panic with bare strings or `fmt.Errorf`, so genuine data corruption gets tagged `reason="other"`. 
For example `generic.go:594`, a custom-bucket index out of range, reachable from MQE via the bucket iterators in `operators/functions/native_histograms.go`.

Maybe upstream could be changed?

Perhaps we change `other` to `unclassified`?
Then reword that bullet to say the panic may be invalid stored data or an engine bug, with the stack trace as the way to tell.

File: operations/mimir-mixin/alerts/alerts.libsonnet [hunk 5]
```diff
@@ -368,6 +368,39 @@ local utils = import 'mixin-utils/utils.libsonnet';
...
+          },
+        },
+        {
+          alert: $.alertName('QueryEngineEvaluationPanics'),
+          expr: |||
+            sum by (%(alert_aggregation_labels)s, reason) (
+              rate(cortex_mimir_query_engine_evaluation_panics_total{reason!="invalid_data"}[%(rate_interval)s])
+            ) > 0
+          ||| % ($._config { rate_interval: $.rateInterval('5m') }),
+          'for': '15m',
```
Comment (karlskewes) [inline operations/mimir-mixin/alerts/alerts.libsonnet:396]: If the panic only happens once, then it will show in the first 5m interval, but it won't `for: 15m` and thus I don't think this alert will fire?

Alternatively would an `increase(...) > 0` expression without a `for:` work here?

WDYT?

**Review by karlskewes — CHANGES_REQUESTED** [review]
